### PR TITLE
Guided Tours: Kill component state, use state.ui.guidesTour

### PR DESCRIPTION
**Pull request by mcsf**
- Note use of createSelector in order to benefit from strict equality
  checks in GuidesTours#shouldComponentUpdate.

- This architecture should make a number of things easier in the future,
  including adding a preview of the frontend during the tour (#4461).

cc @lsinger @ehg

### CODE REVIEW

File: client/state/action-types.js [hunk 1]
```diff
@@ -109,7 +109,9 @@ export const SERVER_DESERIALIZE = 'SERVER_DESERIALIZE';
 export const SET_EXPORT_POST_TYPE = 'SET_EXPORT_POST_TYPE';
 export const SET_ROUTE = 'SET_ROUTE';
 export const SET_SECTION = 'SET_SECTION';
+// TODO(mcsf): CHANGE WORD ORDER OF THESE TWO:
```
Comment (lsinger) [inline client/state/action-types.js]: what / why unclear

Comment (mcsf) [inline client/state/action-types.js]: Adopt the same schema as other action types, where the object precedes the verb: `GUIDESTOUR_SHOW`, `~_UPDATE`. These can be changed when we change the name to `guided-tours`.

File: client/state/ui/selectors.js [hunk 2]
```diff
@@ -44,3 +46,24 @@ export function getSelectedSiteId( state ) {
...
+ *
+ * @param  {Object}  state Global state tree
+ * @return {Object}        Current Guided Tours state
+ */
+const getRawGuidesTourState = state => get( state, 'ui.guidesTour', false );
+export const getGuidesTourState = createSelector(
+	state => {
+		const tourState = getRawGuidesTourState( state );
+		const { stepName = '' } = tourState;
+		const stepConfig = guidesToursConfig[ stepName ] || false;
```
Comment (ehg) [inline client/state/ui/selectors.js]: Should we return `null` if no step is found?

Comment (mcsf) [inline client/state/ui/selectors.js]: I like `false` as it allows property access without throwing an error.

File: client/state/action-types.js (hunk 1 above)
Comment (lsinger) [inline client/state/action-types.js]: ah, gotcha. why not in this PR?

Comment (mcsf) [inline client/state/action-types.js]: I wanted to keep the impact of this PR minimal, and it seemed to make
sense to do it in a PR where we change the naming in general.

**Comment by ehg** [comment]
LGTM, although the tests aren't passing?

We may want to consider moving into our own `ui/guidedtours` folder soon.

**Comment by lsinger** [comment]
> LGTM, although the tests aren't passing?

Something something i18n?

**Comment by ehg** [comment]
> Something something i18n?

Yep, we need an `i18n.initialize()` in the ui selectors test file. Also we should write a selector test for this.

**Comment by mcsf** [comment]
> Yep, we need an i18n.initialize() in the ui selectors test file.

Tried that, did it in `config.js` instead: ad39990.

**Comment by mcsf** [comment]
It could've been done in `selectors.js`, but that seemed more fragile in the long run, and also required some disfiguration of the file (`import` statements seem to be hoisted, so we'd have to convert one of those to `require`, so that we can call `i18n.initialize` after importing `i18n` and before requiring `selectors`.

**Comment by ehg** [comment]
Right, but isn't this just a problem with the tests? By `selectors.js`, d'you mean https://github.com/Automattic/wp-calypso/blob/master/client/state/ui/test/selectors.js? I meant to add the `initialize` there, as well as a test for our new selector.

**Comment by mcsf** [comment]
Yep, I meant `test/selectors.js`.

**Comment by mcsf** [comment]
> but isn't this just a problem with the tests?

It is, in the sense that this never happens in the rest of Calypso because we go through `boot`. But it can happen for _anything_ that happens to require `guidestour/config`, which prompted me to initialize within `config` itself.

Do you think I should avoid that?

**Comment by ehg** [comment]
> Do you think I should avoid that?

Yes, since we handle it both client and server, and use this for the tests: https://github.com/Automattic/wp-calypso/tree/master/test/test/helpers/use-i18n :)

**Comment by mcsf** [comment]
> use this for the tests

TIL :/

I'll get that in, then, thanks!

**Comment by mcsf** [comment]
Actually, it doesn't work: there's the same issue as the one I [mentioned here](https://github.com/Automattic/wp-calypso/pull/4661#issuecomment-209545154), with the added difficulty that `useI18n` needs to live inside a `describe`. See train wreck:

```diff
diff --git a/client/state/ui/test/selectors.js b/client/state/ui/test/selectors.js
index 628ba14..e91f167 100644
--- a/client/state/ui/test/selectors.js
+++ b/client/state/ui/test/selectors.js
@@ -6,7 +6,9 @@ import { expect } from 'chai';
 /**
  * Internal dependencies
  */
-import { getSelectedSite, getSelectedSiteId, getSectionName } from '../selectors';
+import useI18n from 'test/helpers/use-i18n';
+useI18n();
+const { getSelectedSite, getSelectedSiteId, getSectionName } = require( '../selectors' );
 
 describe( 'selectors', () => {
 	describe( '#getSelectedSite()', () => {
```

Sooo: we either do something like in the diff above, but with `i18n.initialize` proper, or we go the other way (ad39990).

Other scenario: somehow rewrite `guidestour/config.js` in a way that `i18n.translate` is never called at require time. :/

**Comment by mcsf** [comment]
My previous comment was pretty confusing and missing the main point, which is: the error happens not during test time proper (i.e. inside a `describe` or `it` statement), but rather at eval time for module `test/selectors` because it's during the requiring `guidestours/config` (happening in `selectors`) that the error happens (at `i18n.translate`).

Which is why we either prevent that at module-eval-time in `test/selectors`, *or* in `guidestours/config` itself.

**Comment by ehg** [comment]
Ah ha. Interesting. A common pattern is to `require` the functions you want to test inside a before block. That should fix it, no?

I do like the idea of maybe a `.get()` function for config, as it could allow us to inject dependencies like i18n, when we want to auto-gen tour READMEs. Let's leave it as is for now though ;)

**Comment by ehg** [comment]
> I do like the idea of maybe a .get() function for config, as it could allow us to inject dependencies like i18n, when we want to auto-gen tour READMEs. Let's leave it as is for now though ;)

After saying all that, I ended up committing 2df85fe

**Comment by mcsf** [comment]
Cool, though could we cache the result of .get() somewhere? Doesn't even have to be in config.js—could be in the selectors file. 

For instance, using createSelector would be a way to define a memoized getter for the config that could care only about certain dependants (I.e. if these change, then call config.get() again).

**Comment by ehg** [comment]
> Cool, though could we cache the result of .get() somewhere? Doesn't even have to be in config.js—could be in the selectors file.

Go ahead ;) Could we add a test too? Then I think we're good to go.

**Comment by mcsf** [comment]
Pushed memoization and tests.

**Comment by ehg** [comment]
LGTM

**Pull request merged**